### ISSUE TITLE
provide conmon documentation

### ISSUE THREAD

**Issue by opuk** [containers/libpod#2770]
**Description**

Currently this is very little to none documentation available on conmon and what it actually does.

There are some vague statements at https://cri-o.io/ for example, such as "Each container is monitored by a separate conmon process. The conmon process holds the pty of the PID1 of the container process. It handles logging for the container and records the exit code for the container process."

This doesn't make it clear on what function it actually does. 
To avoid confusion it would be helpful if there was some documentation available on it's function and what it means that "each container is monitored by a separate conmon process" for example. Can conmon take any action if it detects some error with the container?

What does "handles logging" mean?
The architecture image on https://cri-o.io/ is a good starting point but perhaps it can be better described what it then integrates with, when I quickly look through the code I can see journald for example being used. 
This type of integration would be good to have documented so that I can better understand what it does.

**Comment by mheon** [comment]
We haven't added much documentation on Conmon because we treat it purely as an implementation detail; no one should be using Conmon except for Podman and CRI-O, and we don't even put it in the system's default $PATH as such.

In brief, Conmon is an extremely small C program with almost no memory footprint that wraps the OCI runtime (usually `runc`) that runs the actual container. Upon being launched by Podman, it double-forks to daemonize and detach from the parent that launched it, then launches the OCI runtime as its child. It will live for as long as the runtime (and container) live, and when the container dies it records its exit time and exit code for Podman or CRI-O. While the container is alive, Conmon provides a socket for attaching to the container, holding open the container's standard streams and forwarding them over the socket (and also writes the contents of the container's streams to a log file so `podman logs` and CRI-O can read them). And that's about it.

In short, it's an extremely minimal C binary that manages containers so that Podman processes can die in the foreground and the container will still be under our supervision, available to be attached to and recording exit information when it dies. Each container has a single Conmon process that acts as its parent until the container dies, and each Conmon process manages a single container.

**Comment by mheon** [comment]
A further note: "can conmon take any action if it detects some error with the container?"

No. Conmon is extremely minimal, and doesn't have the logic to detect or handle errors. If we wanted to add such logic, we'd do it in Podman and CRI-O, to keep Conmon simple and its memory footprint minimal.

**Comment by rhatdan** [comment]
@mheon What you wrote above should probably be added to the conmon readme. :^)

**Comment by vrothberg** [comment]
I agree. I asked @opuk on IRC to open the issue here.

Although conmon does not do much, the more curious users (and devs) are currently having a hard time finding out what conmon does. For some, it might seem suspicious that some additional but undocumented binary is installed.

**Comment by rhatdan** [comment]
We should also create a man page for it.

**Comment by baude** [comment]
do folks agree this should be moved to the conmon repo?

**Comment by haircommander** [comment]
Actually conmon has documentation now at `github.com/containers/conmon`. Closing this issue as such, reopen over there if you disagree
